### ISSUE TITLE
Replace usage of has_key() in twisted.mail

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6486]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#6486|
|Type|task|
|Created|2013-04-27 18:58:37Z|
|Branch|https://github.com/twisted/twisted/tree/has_key-6486-2|

The following modules in `twisted.mail` contain calls to the deprecated `x.has_key(y)` method. They should be replaced with `y in x` instead. They're all covered by tests already, except for `twisted.mail.pb` which also doesn't show up in the coverage report?

```
./twisted/mail/mail.py:36:    def has_key(self, name):
./twisted/mail/pb.py:102:        if not self.domains.has_key(domain):
./twisted/mail/pb.py:105:        if (domain.dbm.has_key(name) and
./twisted/mail/test/test_mail.py:84:            self.failUnless(d.has_key(x))
./twisted/mail/test/test_mail.py:723:         self.failUnless(m.has_key('Received'))
./twisted/mail/maildir.py:480:        if not self.dbm.has_key(name):
```

Attachments:

* [my-twisted-patch.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/081/0811812f1a106d6206f32a8c5ed0fde32d27a1b8/d5bf14bfe4ce8fc7b9256a069dbe396aeaef28e5.patch) (2750 bytes) - added by mulhern on 2013-05-02 01:30:49Z - 
* [my-twisted-patch.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/081/0811812f1a106d6206f32a8c5ed0fde32d27a1b8/5d15195804101e72187ade5d2518dfe82783ad6d.patch) (1781 bytes) - added by mulhern on 2013-06-03 15:22:44Z - New patch
* [has_key_6486.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/081/0811812f1a106d6206f32a8c5ed0fde32d27a1b8/383ee0db4c7f424fae7e0543d272b2defed8137a.patch) (2800 bytes) - added by lbromo on 2013-09-23 13:13:56Z - Added documentation

<details><summary>Searchable metadata</summary>

```
trac-id__6486 6486
type__task task
reporter__thijs thijs
priority__low low
milestone__Python_3_x Python-3.x
branch__branches_has_key_6486_2 branches/has_key-6486-2
branch_author__mulhern__lvh__rwall mulhern, lvh, rwall
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__easy easy
time__1367089117000000 1367089117000000
changetime__1380024079000000 1380024079000000
version__None None
owner__rwall rwall
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="mulhern's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|mulhern commented|
|-|-|

The occurrence of has_key in mail/mail.py : 36 is a definition of a class method for the class DomainWithDefaultDict.

DomainWithDefaultDict also contains an __contains__ method (implementing in).

Should the definition of the has_key method in DomainWithDefaultDict be removed? or does it constitute part of a public interface?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

Anything that is a public (doesn't start with underscore) method of a public class cannot be removed or otherwise made non-backwards compatible unless it's been deprecated for more than a year and two releases.

**Comment by twisted-trac** [comment]
|<img alt="mulhern's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|mulhern set owner to mulhern<br>mulhern set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="mulhern's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|mulhern removed owner<br>mulhern set status to `new`|
|-|-|

Removed the class MaildirBroker In twisted/mail/pb.py because it appears to be broken and obsolete. Here are the reasons:
1) No unit tests.
2) Last change to the file was two years ago.
3) Most recent update to code in this class is revision 581.
4) Running either of the two methods getCollections or proto_getCollections on a MaildirBroker object results in an AttributeError. The attribute errors indicate that the field "domains" or the field "_getCollection" do not exist in the object.
5) No occurrences of the name MaildirBroker exist anywhere else in the codebase.

Did not change test_mail.py:84 since it is there explicitly to test the has_key function.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)|@lvh set owner to @lvh|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_6"></a>|@lvh set owner to mulhern|
|-|-|

Hi mulhern,

Thanks again for your contribution :)

Unfortunately, the [compatibility policy](https://twistedmatrix.com/trac/wiki/CompatibilityPolicy) prevents us from yanking it out, even if it's broken and despite all the issues with it you highlighted.

If MaildirBroker is busted, file a ticket for it. According to the compatibility policy, if it's an egregious failure of specification, it can be fixed immediately. If it's obsolete (why?) and needs to be removed, since it's public API, that's the 1 year and 2 release minimum there (as itamar mentioned).

Otherwise, this patch looks okay -- reinstate MaildirBroker and I think we're done :)

thanks again,
lvh

**Comment by twisted-trac** [comment]
|<img alt="mulhern's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|mulhern set owner to @lvh|
|-|-|

New patch omits change to twisted.mail.pb.py and new ticket ([#6547](https://github.com/twisted/twisted/issues/6547)) submitted regarding MaildirBroker's presumed obsolescence.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_8"></a>|@lvh commented|
|-|-|

(In [38643]) Branching to 'has_key-6486'

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_9"></a>|@lvh commented|
|-|-|

Thanks again, mulhern :)

Waiting on buildbots: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/has_key-6486

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_10"></a>|@lvh commented|
|-|-|

There's a remaining issue with your patch:

```
        self.assertIsNone(self.D.userDirectory('nouser')) 
```

assertIsNone isn't a thing: you probably mean:

```
        self.assertIdentical(self.D.userDirectory('nouser'), None) 
```

Since I missed it in the first review, and it's obvious what you meant, I'll commit the appropriate thing. Unfortunately that means I can't be the next reviewer :)

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_11"></a>|@lvh removed owner|
|-|-|

Last patch was committed to the branch and then I added two commits: r38648, r38649. State of the branch is what's up for review.

Buildbot:
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/has_key-6486

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra set owner to mulhern|
|-|-|

Replying to [lvh](#note_11):
> Last patch was committed to the branch and then I added two commits: r38648, r38649. State of the branch is what's up for review.

Thanks. There are a few new twistedchecker [errors](http://buildbot.twistedmatrix.com/builders/twistedchecker/builds/752/steps/run-twistedchecker/logs/new%20twistedchecker%20errors), assigning back to mulhern.

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|Alex commented|
|-|-|

On `twisted/mail/maildir.py:L480`, I think you want to replace `not name in self.dbm` with `name not in self.dbm`

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_15"></a>|@wallrj commented|
|-|-|

(In [40068]) Apply has_key_6486.patch from lbromo. refs [#6486](https://github.com/twisted/twisted/issues/6486)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_16"></a>|@wallrj commented|
|-|-|

(In [40079]) Branching to 'has_key-6486-2'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_17"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Hi lbromo,

* I applied your patch.
* Addressed AlexGaynor's comment.
* Modified one or two of the new docstrings
* Added a newsfile.
* Merged forward.

The build results look good. 

So I'll merge it. Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_18"></a>|@wallrj set status to `closed`|
|-|-|

(In [40095]) Merge has_key-6486-2

Authors: mulhern, lbromo
Reviewers: lvh, rwall, thijs, Alex
Fixes: [#6486](https://github.com/twisted/twisted/issues/6486)

Remove use of has_key in twisted.mail.maildir and twisted.mail.test.test_mail.
Added some missing test docstrings in twisted.mail.test.test_mail.
